### PR TITLE
feat: new cli plugin

### PR DESCRIPTION
**Pull request by caohuilin**
### New CLI Plugin Define

- new context and hook api define
- change use hookRunner to `api.xxx` to define hooks
- add new plugin manager to handle use plugin order
- support modify rsbuild config hooks

## Checklist

- [x] I have added changeset via `pnpm run change`.
- [ ] I have updated the documentation.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: packages/toolkit/plugin-v2/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+export { createPluginManager } from './manager';
+export { createAsyncHook, createCollectAsyncHook } from './hooks';
+export * from './cli';
+
+export * from './types';
```
Comment (zllkjc) [inline packages/toolkit/plugin-v2/src/index.ts:5]: 同，不建议 `export *`

File: packages/toolkit/plugin-v2/src/hooks.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    callbacks.push(cb);
+  };
+
+  const call = async (...params: Parameters<Callback>) => {
+    const results: ReturnType<Callback>[] = [];
+    for (const callback of callbacks) {
+      const result = await callback(...params);
+
+      if (result !== undefined) {
+        params[0] = result;
```
Comment (zllkjc) [inline packages/toolkit/plugin-v2/src/hooks.ts:45]: 这里 params 一直更新是符合预期的吗，目前看的确没有「需要入参」的 CollectAsyncHook

File: packages/solutions/app-tools/src/old.ts [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        filename: string;
+        eventType: string;
+        isPrivate: boolean;
+      }) {
+        const { filename, eventType, isPrivate } = e;
+
+        if (!isPrivate && (eventType === 'change' || eventType === 'unlink')) {
+          const { closeServer } = await import('./utils/createServer.js');
+          await closeServer();
+          await restart(api.useHookRunners(), filename);
```
Comment (zllkjc) [inline packages/solutions/app-tools/src/old.ts:182]: restart will invoke cli.init, we should migrate to new cli instance, now it's in `@modern-js/core`

File: packages/toolkit/plugin-v2/src/hooks.ts (hunk 2 above)
Comment (caohuilin) [reply packages/toolkit/plugin-v2/src/hooks.ts:45]: 改成没有参数吧，后续需要了再提供需要参数的 Hook

File: packages/toolkit/plugin-v2/src/cli/run/create.ts [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+          }
+
+          process.nextTick(() => {
+            process.exit(hasError ? 1 : 0);
+          });
+        });
+      },
+    );
+
+    const extraConfigs = await context.hooks.config.call();
```
[the remark this answers is not in the document]
Comment (caohuilin) [reply packages/toolkit/plugin-v2/src/cli/run/create.ts:121]: beforeConfig 的内容在新的 app tools setup 中实现了，这里感觉应该是不需要了

File: packages/toolkit/plugin-v2/src/cli/api.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+          extendsPluginApi[apiName] = apis[apiName];
+        });
+      }
+    });
+  }
+
+  updateRegistryApi();
+
+  (Object.keys(extendsHooks) as ExtendsHooksKey[]).forEach(hookName => {
+    extendsPluginApi[hookName] = extendsHooks[hookName].tap;
```
[the remark this answers is not in the document]
Comment (caohuilin) [reply packages/toolkit/plugin-v2/src/cli/api.ts:65]: 后续看看能不能实现

**Review by zllkjc — APPROVED** [review]

**Pull request merged**